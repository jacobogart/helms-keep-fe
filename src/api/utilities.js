export const host = 'http://localhost:5000/api/';

export const mockTask = {
  id: '1234',
  isCompleted: false, 
  message: 'Clean the kitchen'
};

export const mockCompletedTask = {
  id: '5678',
  isCompleted: true,
  message: 'Make a mock completed task'
}

export const mockQuest = {
	id: '1234',
  title: 'My First Quest',
  color: "green",
	challenges: [ mockTask, mockCompletedTask ]
};
