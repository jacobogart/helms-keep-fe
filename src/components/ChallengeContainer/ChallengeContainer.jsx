import React, { Component } from 'react';
import Challenge from '../../containers/Challenge/Challenge';
import shortid from 'shortid';
import PropTypes from 'prop-types';
export class ChallengeContainer extends Component {
	state = {
		chalInput: ''
	};

	// componentDidUpdate() {
	// 	this.scrollToBottom();
	// }

	// scrollToBottom() {
	// 	this.el.scrollIntoView({ behavior: 'smooth' });
	// }

	handleChange = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	handleSubmit = e => {
		e.preventDefault();
    const challenge = { id: shortid.generate(), message: this.state.chalInput, isCompleted: false };
    const { viewType } = this.props;

		if (viewType === 'list') {
			this.props.addChallenge(challenge);
		} else {
			this.props.saveChallenge(challenge);
		}
		this.setState({ chalInput: '' });
	};

  render() {
		const ulClass = this.props.challengeType === 'complete' ? 'complete-ul' : 'incomplete-ul';

		const newChallengeInput =
			this.props.challengeType !== 'complete' ? (
        <form className="form-row" onSubmit={this.handleSubmit}>
					<input
						className="new-challenge-input"
						placeholder="+ Add new challenge..."
						name="chalInput"
						onChange={this.handleChange}
						value={this.state.chalInput}
					/>
					<button type="submit" className="add-challenge-btn" disabled={!this.state.chalInput}>
						+
					</button>
				</form>
			) : null;

		const challenges = this.props.challenges.map(chal => {
			return (
				<Challenge
					data={chal}
					viewType={this.props.viewType}
					key={chal.id}
					updateChallenge={this.props.updateChallenge}
          deleteChallenge={this.props.deleteChallenge}
          editChallenge={this.props.editChallenge}
          removeChallenge={this.props.removeChallenge}
				/>
			);
		});

		return (
			<article className="ChallengeContainer">
				<ul className={ulClass}>
					{challenges}
					<div
						ref={el => {
							this.el = el;
						}}
					/>
				</ul>
				{newChallengeInput}
			</article>
		);
	}
}
ChallengeContainer.propTypes = {
	addChallenge: PropTypes.func,
	challenges: PropTypes.arrayOf(PropTypes.object),
	deleteChallenge: PropTypes.func,
	type: PropTypes.string,
	updateChallenge: PropTypes.func
  };

export default ChallengeContainer;
