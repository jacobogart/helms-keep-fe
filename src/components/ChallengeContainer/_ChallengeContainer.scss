.ChallengeContainer {
  overflow-y: scroll;
  // height: 280px;
} 

.complete-ul {
  span,
  .far,
  .fas  {
    color: rgba(131, 131, 131, 0.728)!important;
  }

  .fa-times:hover {
    color: darkred;
    opacity: .7;
  }
}
.form-row {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
}

.new-challenge-input {
  margin: 0;
  flex: auto;
  height: 40px;
  padding-left: 20px;
  font-size: 14px;

  &::placeholder {
    color: rgb(184, 184, 184);
  }
}

.add-challenge-btn {
  height: 40px;
  border: none;
  padding: 0 15px;
  color: white;
  font-size: 26px;
}
