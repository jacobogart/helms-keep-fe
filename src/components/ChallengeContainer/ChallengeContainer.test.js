import React from 'react';
import { shallow } from 'enzyme';
import ChallengeContainer from './ChallengeContainer';
import { mockTask } from '../../api/utilities';
import shortid from "shortid";

jest.mock("shortid");

describe('ChallengeContainer', () => {
  let wrapper, instance;

  const mockAddChallenge = jest.fn();
  const mockDeleteChallenge = jest.fn();
  const mockUpdateChallenge = jest.fn();
  const mockSaveChallenge = jest.fn();
  const mockChallenges=[mockTask, mockTask];
  const mockChallenge = { id: 1, isCompleted: false, message: "beef" };

  const mockSubmit = { preventDefault: () => {} }
  const mockChange = {
    target: {
      name: 'chalInput',
      value: 'beef'
    }
  }

  shortid.generate.mockImplementation(() => 1);

  beforeEach(() => {
    wrapper = shallow(
      <ChallengeContainer
        addChallenge={mockAddChallenge}
        challenges={mockChallenges}
        deleteChallenge={mockDeleteChallenge}
        saveChallenge={mockSaveChallenge}
        challengeType="incomplete"
        viewType="list"
        updateChallenge={mockUpdateChallenge}
      />,
    );
    
    instance = wrapper.instance();

    instance.el = {
      scrollIntoView: jest.fn()
    }
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when type is complete', () => {
    wrapper = shallow(
      <ChallengeContainer
        challenges={mockChallenges}
        saveChallenge={mockSaveChallenge}
        type="complete"
      />,
      { disableLifecycleMethods: true }
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when type is form', () => {
    wrapper = shallow(
      <ChallengeContainer
        challenges={mockChallenges}
        saveChallenge={mockSaveChallenge}
        type="form"
      />,
      { disableLifecycleMethods: true }
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    const defaultState = {
      chalInput: ""
    };
    expect(wrapper.state()).toEqual(defaultState);
  });

  // it('should invoke scrollToBottom on update', () => {
  //   jest.spyOn(instance, 'scrollToBottom');
  //   instance.componentDidUpdate();
  //   expect(instance.scrollToBottom).toHaveBeenCalled();
  // });

  describe('Event Listeners', () => {
    it('should invoke handleSubmit on submit', () => {
      wrapper.find('.form-row').simulate('submit', mockSubmit);
      expect(mockAddChallenge).toHaveBeenCalled();
    });

    it("should invoke handleChange on input change", () => {
      wrapper.find(".new-challenge-input").simulate("change", mockChange);
      expect(wrapper.state('chalInput')).toEqual('beef');
    });
  });

  // describe('scrollToBottom', () => {
  //   it('should invoke scrollIntoView when called', () => {
  //     instance.scrollToBottom();
  //     expect(instance.el.scrollIntoView).toHaveBeenCalled();
  //   });
  // });

  describe('handleChange', () => {
    it('should change the appropriate state to the appropriate value', () => {
      instance.handleChange(mockChange);
      expect(wrapper.state('chalInput')).toEqual('beef');
    });
  });

  describe('handleSubmit', () => {
    it('should invoke addChallenge if type is incomplete', () => {
      instance.handleChange(mockChange);
      instance.handleSubmit(mockSubmit);
      expect(mockAddChallenge).toHaveBeenCalledWith(mockChallenge);
    });

    it("should invoke saveChallenge if type is not incomplete", () => {
      wrapper = shallow(
        <ChallengeContainer
          challenges={mockChallenges}
          saveChallenge={mockSaveChallenge}
          challengeType="complete"
          viewType="form"
        />,
        { disableLifecycleMethods: true }
      );
      instance = wrapper.instance();
      instance.handleChange(mockChange);
      instance.handleSubmit(mockSubmit);
      expect(mockSaveChallenge).toHaveBeenCalledWith(mockChallenge);
    });

    it('should clear chalInput in state', () => {
      instance.handleChange(mockChange);
      expect(wrapper.state('chalInput')).toEqual('beef');
      instance.handleSubmit(mockSubmit);
      expect(wrapper.state("chalInput")).toEqual("");
    });
  });
});