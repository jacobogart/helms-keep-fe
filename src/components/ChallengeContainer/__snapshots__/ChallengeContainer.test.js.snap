// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengeContainer should match the snapshot 1`] = `
<article
  className="ChallengeContainer"
>
  <ul
    className="incomplete-ul"
  >
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      deleteChallenge={[MockFunction]}
      key="1234"
      updateChallenge={[MockFunction]}
      viewType="list"
    />
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      deleteChallenge={[MockFunction]}
      key="1234"
      updateChallenge={[MockFunction]}
      viewType="list"
    />
    <div />
  </ul>
  <form
    className="form-row"
    onSubmit={[Function]}
  >
    <input
      className="new-challenge-input"
      name="chalInput"
      onChange={[Function]}
      placeholder="+ Add new challenge..."
      value=""
    />
    <button
      className="add-challenge-btn"
      disabled={true}
      type="submit"
    >
      +
    </button>
  </form>
</article>
`;

exports[`ChallengeContainer should match the snapshot when type is complete 1`] = `
<article
  className="ChallengeContainer"
>
  <ul
    className="incomplete-ul"
  >
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      key="1234"
    />
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      key="1234"
    />
    <div />
  </ul>
  <form
    className="form-row"
    onSubmit={[Function]}
  >
    <input
      className="new-challenge-input"
      name="chalInput"
      onChange={[Function]}
      placeholder="+ Add new challenge..."
      value=""
    />
    <button
      className="add-challenge-btn"
      disabled={true}
      type="submit"
    >
      +
    </button>
  </form>
</article>
`;

exports[`ChallengeContainer should match the snapshot when type is form 1`] = `
<article
  className="ChallengeContainer"
>
  <ul
    className="incomplete-ul"
  >
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      key="1234"
    />
    <ConnectFunction
      data={
        Object {
          "id": "1234",
          "isCompleted": false,
          "message": "Clean the kitchen",
        }
      }
      key="1234"
    />
    <div />
  </ul>
  <form
    className="form-row"
    onSubmit={[Function]}
  >
    <input
      className="new-challenge-input"
      name="chalInput"
      onChange={[Function]}
      placeholder="+ Add new challenge..."
      value=""
    />
    <button
      className="add-challenge-btn"
      disabled={true}
      type="submit"
    >
      +
    </button>
  </form>
</article>
`;
